src/datadog/tracer_signature.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+// This component provides a class, `TracerSignature`, that contains the parts
+// of a tracer's configuration that are used to refer to the tracer in Datadog's
+// telemetry and remote configuration APIs.
+//
+// `TracerSignature` is used in three contexts:
+//
+// 1. When telemetry is sent to the Datadog Agent, the tracer signature is
+//    included in the request payload. See
+//    `TracerTelemetry::generate_telemetry_body` in `tracer_telemetry.cpp`.
+// 2. When the Datadog Agent is polled for configuration updates, part of the
+//    tracer signature (all but the language version) is included in the request
+//    payload. See `RemoteConfigurationManager::make_request_payload` in
+//    `remote_config.h`.
+// 3. When the Datadog Agent responds with configuration updates, the service
+//    and environment of the tracer signature are used to determine whether the
+//    updates are relevant to the `Tracer` that created the collector that is
+//    polling the Datadog Agent. See
+//    `RemoteConfigurationManager::process_response` in `remote_config.h`.
+
+#include <string>
+
+#include "runtime_id.h"
+#include "string_view.h"
+
+namespace datadog {
+namespace tracing {
+
+class TracerSignature {
```

Comment:
Does it have to be a `class` with getter?

Reply:
Looking at the two (well, more than two) JSON payloads that use `service`, `environment`, and `runtime_id`, I saw that library language, library version, and library language version (all constants) were also there on an equal footing. Who's to say that those are not also part of the signature?

So, either I would keep it a `struct` having only the three fields, and leave the code that pulls the other values from the appropriate constants; or, I could wrap those constants into the type. I chose the latter, but am not committed to the idea.

I justified it to myself by saying "this makes the notion of a tracer signature more substantial."

Reply:
How about a third option where it's still a `struct` or `class` but with all fields accessible, like:

````cpp
struct TracerSignature {
  RuntimeID runtime_id_;
  std::string default_service_;
  std::string default_environment_;
  StringView library_version = tracer_version;
  StringView library_language = "cpp";
  StringView library_version = __cplusplus;

  TracerSignature() = delete;
  TracerSignature(const RuntimeID& runtime_id, ...) : runtime_id(runtime_id) ... {}
};
````

Reply:
That works.

`__cplusplus` is an integer literal, though.

Options that I considered:

1. Just use `std::string`. That would work.
2. Use the preprocessor to stringify __cplusplus. This works, but then you get a trailing "L" due to the definition of the __cplusplus token. :(
3. Use member functions. :D

It also ever so slightly bothers me that `library_*` as data members implies that the values might mutate, when in actually the never will. Maybe this is an indication that it's not such a good idea to put them in `TraceSignature` after all.

`const` data members aren't worth the trouble, seeing how they disable assignments and moves, though that might not matter for `TracerSignature`'s usage.

Anything will do.

Reply:
Too cute? :P

https://godbolt.org/z/6xbY5c4Mf

Reply:
Perfect. Will replace `substr` with:
````cpp
std::string_view version = std::string_view{STRINGIFY(__cplusplus), 6};
````

Reply:
```c++
std::string_view version = "most likely 201703";
```